Spec: NuGet Audit Sources in Visual Studio Options (#14589)
NuGet/Home: accepted/2025/audit-sources-visual-studio-options.md

#### Read vulnerabilities from my package sources

- Default option - if no audit sources exist, this will be the selection.
- Audit Sources table will be **hidden**.

![Existing package sources Visual Studio Options page shown with a new "Choose how NuGet Audit retrieves vulnerability data" dropdown](../../meta/resources/AuditSources-VS/audit-choose-package-source-page.png)

#### Configure sources to read vulnerabilities

- Audit Sources table will be **shown**.
- **Pre-selected**  when one or more audit sources are already configured.
  - The dropdown will be disabled as well since the presence of audit sources takes away the behavior of reading vulnerability data from package sources.
- **User-selectable** only when no audit sources exist, enabling  a customer to explicitly configure their first `<auditSource>` using the `Add` button.

![New dropdown expanded with both options shown"](../../meta/resources/AuditSources-VS/choose-nuget-audit-dropdown.png)

![Table of Audit Sources shown below Package Sources because Combobox option "Configure sources to read vulnerabilities" is selected](../../meta/resources/AuditSources-VS/audit-sources-visible-table.png)

> - Switching the dropdown back to "**Read vulnerabilities from my package sources**" would not be supported in this iteration.

Reviewer: Why not? Seems straightforward to implement.

Author: NuGet Restore will use audit sources if they exist. 
What isn't straightforward is making the uncheck automatically remove all of those audit sources. 
I don't expect more than an average of 1 audit source, maybe 2-3 audit sources in the more extreme cases. It's also probably less common to go from audit sources back to package source vulnerability resources.

Therefore, customers can press Remove on all their audit sources and end up back in the original state. Additionally, the Unified Settings team is exploring building in a Remove All button on array settings.

Reviewer: In that case, could we simplify the UX by always showing a list of Audit Sources (even if empty)? Having a dropdown that only works for opting in but not for opting out seems a bit broken.

Author: We discussed this option first, and we believe it's confusing to have an empty table taking up space that isn't used by Audit in most cases. We agreed on hiding it entirely behind a user's choice to explicitly configure Audit. This makes it a tangible act that you're trying to manage your audit sources yourself.

It'll be a Checkbox instead of a Dropdown, as mentioned - but it doesn't opt you in to anything. It just enables you to choose audit sources. We do not have any setting today that opts you in to using audit sources; we use them if they are present.

Reviewer 2: I like the part of the spec on the next line that covers how to revert back to using only package sources as the audit sources:
> If customers want the ability to switch back from their audit sources to only package sources for Vulnerability data, a future iteration could support this and automatically clear `<auditSources>` after showing a warning messagebox that can be cancelled.

I think this would be a great addition for the first iteration. Allowing users to clear their audit sources directly in the UI creates a complete experience. Is there a way to programmatically iterate over the list of audit sources and remove each one from the nuget.config file(s) when the user confirms toggles the checkbox and confirms in the dialog?

Reviewer 2: We talked offline and I understand why we want to consider other options for clearing all the audit sources later. My input should be considered resolved and not block this thread but I'm going to rely on the comment author to resolve the comment when ready.

- Use existing NuGet.Configuration APIs to read/write `<auditSources>` in `nuget.config` files.

#### Telemetry

- Track management actions (add/remove/enable) of audit sources to better understand customer needs.

## **Drawbacks**

- Potential confusion for package sources that act as audit sources implicitly by having a vulnerability resource.

## **Rationale and Alternatives**

### Alternative 1: Always show Audit Sources table

No ComboBox to select - the Audit Sources table will always be available below package sources, relying on explanatory text to explain the behavior changes if any are configured.

![Table of Audit Sources shown below Package Sources in the existing package sources Visual Studio Options page](../../meta/resources/AuditSources-VS/audit-source-vs-options.png)

### Alternative 2: Create a Separate Options Page

Rather than share the page with Package Sources, create a new section, **Audit Sources**

- Advantages:

  - Less clutter on the Package Sources page.

- Disadvantages:

  - Separate UI for Package & Audit sources is potentially confusing, or at least more cumbersome for customers, as they are conceptually tightly related.

>   - More duplication of code to setup a separate page which is nearly identical to the Package Sources page.

Reviewer 2: Agreed. Since package sources and audit sources are related in that package sources can act as the audit sources when audit sources are not provided, having them configurable in the same dialog window would present a better experience I think.

## **Explanation**

### **Functional Explanation**

Add an "Audit Sources" table to the "Package Sources" page in NuGet's Visual Studio Options.
This will provide a unified, intuitive UI experience for managing Audit Sources alongside Package Sources within Visual Studio.

Discoverability is further increased by enabling the Quick Search toolbar (Ctrl+Q) where searching "Audit Source" will direct customers into the Package Sources options page in the Unified Settings platform.

#### Package Sources page

- Add a table titled "Audit Sources" to the package sources page which lists the explicitly configured Audit Sources from the `<auditSources>` section.
- "Audit Sources" table will be functionally identical to the "Package Sources" table.
- One table and up to Three tables will now be shown on this page:
  - Package Sources (always shown)
  - Audit Sources (shown when explicitly configured),
  - Machine-wide Package Sources (shown when explicitly configured)

#### Describe Package versus Audit source

Before each table, introduce descriptive text to inform customers of how Package Sources and Audit Sources work together.

- **Package sources**:

>   > Configure the sources NuGet will use to for displaying and downloading packages. NuGet Audit will also reference vulnerability data from sources that support it. Alternatively, dedicated Audit Source(s) can be configured below.

Reviewer 3: I think the message about package sources being used for audit only when no audit sources are configured can be more clear.

```suggestion
  > Configure the sources NuGet will use to for displaying and downloading packages. When there are no Audit Source(s) configured (see below), then package sources will also reference vulnerability data.
```

Author: Totally - these are sample strings, not final strings. I believe we can discuss and have UX/PM review the strings when the spec implementation begins.

- **Audit sources**:

  > Configure the sources NuGet Audit will use for retrieving Package Vulnerability data. If none are configured, any configured package sources that support Vulnerability data will be used by NuGet Audit.

#### Mockup

Unified Settings mockup showing a new Audit Sources table below the existing Package Sources:

![Table of Audit Sources shown below Package Sources in the existing package sources Visual Studio Options page](../../meta/resources/AuditSources-VS/audit-source-vs-options.png)

### **Technical Explanation**

- Add an array setting titled "Audit Sources" to the "Package Sources" NuGet options page in the Unified Settings registration.json file.
- Use existing NuGet.Configuration APIs to read/write `<auditSources>` in `nuget.config` files.

#### Telemetry

- Track management actions (add/remove/enable) of audit sources to better understand customer needs.

## **Drawbacks**

- Potential confusion for package sources that act as audit sources implicitly by having a vulnerability resource.

> ## **Rationale and Alternatives**

Reviewer 3: Another alternative could be to have a single sources list, and have a new column to allow customers to choose whether it's an package source, audit source, or both.  This might make it more important to clearly explain that no audit sources means that audit falls back to package sources.  But the advantage is that it becomes clearer to customers that when you disable a source with a specific name (or URL? I can't remember) it disables both audit and package sources with that name (or url).  We don't (yet?) have a GUI to edit credentials, but if we did, they similar apply to both package and audit sources.

Author: At first, I didn't think this was a possibility, but thinking through your comment, I believe we could implement it that way.
For `Both`, we would have to add the source to `<packageSources>` and to `<auditSources>`, effectively removing the concept of an implicit audit source. Is that your line of thinking as well?

Author: Added as Alternative 3.
We also discussed the main proposal offline, so my understanding is you agree this is an alternative and not the primary proposal.

- A new section, **Audit Sources**, would take the approach of copying the Package Sources options page into a new Audit Sources page, rather than share the same space.

- Advantages:

  - Less clutter on the Package Sources page.

- Disadvantages:

  - Separate UI for Package & Audit sources is potentially confusing, or at least more cumbersome, as they are conceptually tightly related.
  - More duplication of code to setup a separate page which is nearly identical to the Package Sources page.

## **Prior Art**

- Audit Sources are used by NuGet's CLI tooling and recently supported in Visual Studio.
- NuGet.Config support for audit sources was implemented consistently with other settings, so it follows that UI tooling to manage those audit sources would build upon that work.

## **Unresolved Questions**

None at this time

## **Future Possibilities**

1. Consider UI for Package Sources that are implicitly audit sources
    1. For implicit audit sources, modify the Package Sources table to show a checkbox or similar indicator that the package source provides vulnerability data.
    1. Validate audit source URLs by checking for `VulnerabilityInfo` resource in the service index.

> 1. Should there be a button to add `nuget.org` as an audit source?

Reviewer 3: Related, but tangential, is that it would be great to detect customers still using nuget.org's legacy v2 andpoint, and ask them to migrate to the v3 endpoint.

Author: I believe this should be a separately filed issue. Would you like me to file that, or would you like to since you have the context?

Author: https://github.com/NuGet/Home/issues/14614
